### PR TITLE
[ISSUE #905] report data for WAP snapshots when currentSnapshot is null

### PR DESCRIPTION
**Pull request by XiaokunDing**
`all_data_files` ,`all_manifests ` and `all_entries` metadata tables report data for WAP snapshots when currentSnapshot is null.

### CODE REVIEW

**Comment by rdblue** [comment]
I don't think this is a good way to fix the problem. These tables do not need to use special handling for WAP, and should not have anything that depends on WAP.

The underlying problem is that `planFiles` is shared between data and metadata tables, and returns no tasks if there is no current snapshot. The solution is to fix `planFiles` for metadata tables, not to add a special case that incorrectly reports the current snapshot.

**Comment by XiaokunDing** [comment]
Thanks @rdblue , I have fixed it, would mind to take another look?

File: spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java [hunk 1]
```diff
@@ -342,7 +345,7 @@ public void testAllDataFilesTable() throws Exception {
         .save(loadLocation(tableIdentifier));
 
     // delete the first file to test that not only live files are listed
-    table.newDelete().deleteFromRowFilter(Expressions.equal("id", 1)).commit();
+ //   table.newDelete().deleteFromRowFilter(Expressions.equal("id", 1)).commit(); //这里会影响WAP 表
```
Comment (chenjunjiedada) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:348]: Why comment this out?

File: core/src/main/java/org/apache/iceberg/MetadataTableScan.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+  protected MetadataTableScan(
+      TableOperations ops, Table table, Long snapshotId, Schema schema, Expression rowFilter,
+      boolean caseSensitive, boolean colStats, Collection<String> selectedColumns,
+      ImmutableMap<String, String> options) {
+    super(ops, table, snapshotId, schema, rowFilter, caseSensitive, colStats, selectedColumns, options);
+  }
+
+  @Override
+  protected abstract long targetSplitSize(TableOperations ops);
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetadataTableScan.java:46]: Why do we need to duplicate these abstract methods here too?

File: core/src/main/java/org/apache/iceberg/MetadataTableScan.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import com.google.common.collect.ImmutableMap;
+import java.util.Collection;
+import org.apache.iceberg.events.Listeners;
+import org.apache.iceberg.events.ScanEvent;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.io.CloseableIterable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class MetadataTableScan extends BaseTableScan {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetadataTableScan.java:31]: The name `MetadataTableScan` is very generic but it is actually used only in aggregated metadata tables like `all_data_files` and `all_manifests`. I think we either need to change the naming or just override `planFiles()` in each individually. We can, for example, simply check if there is any snapshot at all and if yes, call an overloaded version of `planFiles`.

File: core/src/main/java/org/apache/iceberg/MetadataTableScan.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      TableOperations ops, Table table, Long snapshotId, Schema schema, Expression rowFilter, boolean caseSensitive,
+      boolean colStats, Collection<String> selectedColumns, ImmutableMap<String, String> options);
+
+  @Override
+  protected abstract CloseableIterable<FileScanTask> planFiles(
+      TableOperations ops, Snapshot snapshot, Expression rowFilter, boolean caseSensitive, boolean colStats);
+
+  @Override
+  public CloseableIterable<FileScanTask> planFiles() {
+    Snapshot snapshot = snapshot();
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetadataTableScan.java:59]: We don't have to analyze the current snapshot for `all_data_files` and `all_manifests`.

**Comment by aokolnychyi** [comment]
I think we will also need a test that can be added to `TestIcebergSourceTablesBase`. It should start with an empty table, stage a snapshot or two and verify the records returned by `all_data_files` and `all_manifests`.

**Comment by XiaokunDing** [comment]
Thanks @aokolnychyi , I have fixed it, would mind to take another look?

File: spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java [hunk 5]
```diff
@@ -326,6 +326,50 @@ public void testFilesUnpartitionedTable() throws Exception {
     TestHelpers.assertEqualsSafe(filesTable.schema().asStruct(), expected.get(0), actual.get(0));
   }
 
+  @Test
+  public void testAggregatedMetadataTaleInStage() throws Exception {
```
Comment (rdblue) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:330]: `Tale` -> `Table`?

File: core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import com.google.common.collect.ImmutableMap;
+import java.util.Collection;
+import org.apache.iceberg.events.Listeners;
+import org.apache.iceberg.events.ScanEvent;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.io.CloseableIterable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AggregatedMetadataTableScan extends BaseTableScan {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java:31]: Why is this public?

File: core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(AggregatedMetadataTableScan.class);
+
+  public AggregatedMetadataTableScan(TableOperations ops, Table table, Schema fileSchema) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java:34]: This can be protected or package-private.

File: core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java (hunk 6 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java:31]: These tables aren't aggregates. In fact, some of them produce duplicates (like all_manifests) when a file appears twice. Can we use a better name here? What about `BaseAllMetadataTableScan`?

File: core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  protected AggregatedMetadataTableScan(
+      TableOperations ops, Table table, Long snapshotId, Schema schema, Expression rowFilter,
+      boolean caseSensitive, boolean colStats, Collection<String> selectedColumns,
+      ImmutableMap<String, String> options) {
+    super(ops, table, snapshotId, schema, rowFilter, caseSensitive, colStats, selectedColumns, options);
+  }
+
+  @Override
+  public CloseableIterable<FileScanTask> planFiles() {
+    LOG.info("Scanning table {} in all snapshots with filter {}", table(), filter());
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java:47]: I don't think it's necessary to use "in all snapshots" because a snapshot doesn't make sense for these tables. It's not that this is scanning the `manifests` table across all snapshots, it is scanning the `all_manifests` table that is not specific to a snapshot by definition. Let's change this to `Scanning metadata table {} with filter {}`.

File: core/src/main/java/org/apache/iceberg/BaseTableScan.java [hunk 9]
```diff
@@ -318,7 +318,7 @@ private Schema lazyColumnProjection() {
     return schema;
   }
 
-  private static String formatTimestampMillis(long millis) {
+  protected static String formatTimestampMillis(long millis) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTableScan.java:321]: This isn't used. Could you revert this change?

**Comment by XiaokunDing** [comment]
HI @rdblue I have fixed it, would mind to take another look?

File: core/src/main/java/org/apache/iceberg/BaseTableScan.java (hunk 9 above)
Comment (XiaokunDing) [reply core/src/main/java/org/apache/iceberg/BaseTableScan.java]: OK

File: core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java (hunk 8 above)
Comment (XiaokunDing) [reply core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java]: fixed

File: core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java (hunk 7 above)
Comment (XiaokunDing) [reply core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java]: OK

File: core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java (hunk 6 above)
Comment (XiaokunDing) [reply core/src/main/java/org/apache/iceberg/AggregatedMetadataTableScan.java]: fixed

File: spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java (hunk 5 above)
Comment (XiaokunDing) [reply spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java]: fixed

File: spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java [hunk 10]
```diff
@@ -326,6 +326,50 @@ public void testFilesUnpartitionedTable() throws Exception {
     TestHelpers.assertEqualsSafe(filesTable.schema().asStruct(), expected.get(0), actual.get(0));
   }
 
+  @Test
+  public void testAggregatedMetadataTableInStage() throws Exception {
```
Comment (rdblue) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:330]: This test name is incorrect. Can you please update it to `testAllMetadataTablesWithStagedCommits()`?

Comment (XiaokunDing) [reply spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java]: OK

**Comment by rdblue** [comment]
Thanks, @XiaokunDing! I'm merging this.

**Comment by XiaokunDing** [comment]
Thanks @rdblue
